### PR TITLE
Make Jenkins_Home overwritable

### PR DESCRIPTION
**Pull request by waveywaves**
- Able to set JENKINS_HOME env variable

### CODE REVIEW

**Comment by waveywaves** [comment]
fixes https://github.com/jenkinsci/kubernetes-operator/issues/166

**Comment by waveywaves** [comment]
@tomaszsek @pawelprazak Can you PTAL ?

File: pkg/controller/jenkins/configuration/base/resources/pod.go [hunk 1]
```diff
@@ -79,6 +79,16 @@ func GetJenkinsMasterContainerBaseEnvs(jenkins *v1alpha2.Jenkins) []corev1.EnvVa
 	return envVars
 }
 
+// GetJenkinsHomePath fetches the Home Path for Jenkins
+func GetJenkinsHomePath(jenkins *v1alpha2.Jenkins) string {
+	for _, envVar := range jenkins.Spec.Master.Containers[0].Env {
+		if envVar.Name == "JENKINS_HOME" {
+			return envVar.Value
```
Comment (pawelprazak) [inline pkg/controller/jenkins/configuration/base/resources/pod.go]: do we need to sanitize this input form envs? like, trim whitespace

Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/resources/pod.go]: No we don't need to AFAIK

File: pkg/controller/jenkins/configuration/base/resources/init_configuration_configmap.go [hunk 2]
```diff
@@ -66,7 +66,7 @@ func GetInitConfigurationConfigMapName(jenkins *v1alpha2.Jenkins) string {
 func NewInitConfigurationConfigMap(meta metav1.ObjectMeta, jenkins *v1alpha2.Jenkins) (*corev1.ConfigMap, error) {
 	meta.Name = GetInitConfigurationConfigMapName(jenkins)
 
-	createJenkinsOperatorUserGroovy, err := buildCreateJenkinsOperatorUserGroovyScript()
+	createJenkinsOperatorUserGroovy, err := buildCreateJenkinsOperatorUserGroovyScript(jenkins)
```
[the remark this answers is not in the document]
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/resources/init_configuration_configmap.go]: This should be alright as the output seems to be serialized and deserializing and making changes based on a struct would be a more expensive operation.

File: pkg/controller/jenkins/configuration/base/resources/pod.go [hunk 3]
```diff
@@ -65,7 +65,7 @@ func GetJenkinsMasterContainerBaseEnvs(jenkins *v1alpha2.Jenkins) []corev1.EnvVa
 	envVars := []corev1.EnvVar{
 		{
 			Name:  "JENKINS_HOME",
-			Value: jenkinsHomePath,
+			Value: GetJenkinsHomePath(jenkins),
 		},
 	}
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/base/resources/pod.go]: Please delete those lines and add JENKINS_HOME in `NewJenkinsMasterContainer` function.

File: pkg/controller/jenkins/configuration/base/resources/pod.go [hunk 4]
```diff
@@ -79,6 +79,16 @@ func GetJenkinsMasterContainerBaseEnvs(jenkins *v1alpha2.Jenkins) []corev1.EnvVa
 	return envVars
 }
 
+// GetJenkinsHomePath fetches the Home Path for Jenkins
+func GetJenkinsHomePath(jenkins *v1alpha2.Jenkins) string {
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/base/resources/pod.go]: can be private `getJenkinsHomePath`

File: pkg/controller/jenkins/configuration/base/resources/pod.go [hunk 5]
```diff
@@ -79,6 +79,16 @@ func GetJenkinsMasterContainerBaseEnvs(jenkins *v1alpha2.Jenkins) []corev1.EnvVa
...
+	for _, envVar := range jenkins.Spec.Master.Containers[0].Env {
+		if envVar.Name == "JENKINS_HOME" {
+			return envVar.Value
+		}
+	}
+	return jenkinsHomePath
```
[the remark this answers is not in the document]
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/resources/pod.go]: isn't `return "/var/jenkins/home"` the same as `return jenkinsHomePath`

File: pkg/controller/jenkins/configuration/base/resources/pod.go (hunk 4 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/resources/pod.go]: Done.

File: pkg/controller/jenkins/configuration/base/resources/pod.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomaszsek) [reply pkg/controller/jenkins/configuration/base/resources/pod.go]: Yes it is. If we will remove `jenkinsHomePath` const we force to use the `GetJenkinsHomePath` function.

File: pkg/controller/jenkins/configuration/base/resources/pod.go (hunk 3 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/resources/pod.go]: The validation test on this fails as the unit test is being run against BaseEnvs

Comment (tomaszsek) [reply pkg/controller/jenkins/configuration/base/resources/pod.go]: So we now allow to override the `JENKINS_HOME` the test `override JENKINS_HOME env` can be deleted.

File: pkg/controller/jenkins/configuration/base/resources/pod.go [hunk 6]
```diff
@@ -65,7 +65,7 @@ func GetJenkinsMasterContainerBaseEnvs(jenkins *v1alpha2.Jenkins) []corev1.EnvVa
 	envVars := []corev1.EnvVar{
 		{
 			Name:  "JENKINS_HOME",
-			Value: jenkinsHomePath,
+			Value: getJenkinsHomePath(jenkins),
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/base/resources/pod.go]: Please delete those lines and add JENKINS_HOME in NewJenkinsMasterContainer function.

File: pkg/controller/jenkins/configuration/base/resources/pod.go (hunk 3 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/resources/pod.go]: Okay i will remove the validation accordingly.

File: pkg/controller/jenkins/configuration/base/resources/pod.go [hunk 7]
```diff
@@ -204,7 +209,11 @@ func GetJenkinsMasterContainerBaseVolumeMounts(jenkins *v1alpha2.Jenkins) []core
 // NewJenkinsMasterContainer returns Jenkins master Kubernetes container
 func NewJenkinsMasterContainer(jenkins *v1alpha2.Jenkins) corev1.Container {
 	jenkinsContainer := jenkins.Spec.Master.Containers[0]
-	envs := GetJenkinsMasterContainerBaseEnvs(jenkins)
+	jenkinsHomeEnvVar := corev1.EnvVar{
+		Name:  "JENKINS_HOME",
+		Value: getJenkinsHomePath(jenkins),
+	}
+	envs := append(GetJenkinsMasterContainerBaseEnvs(jenkins), jenkinsHomeEnvVar)
 	envs = append(envs, jenkinsContainer.Env...)
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/base/resources/pod.go]: `jenkinsHomeEnvVar` should be added only if it is not set in `jenkinsContainer.Env`

File: pkg/controller/jenkins/configuration/base/resources/pod.go [hunk 8]
```diff
@@ -204,9 +209,26 @@ func GetJenkinsMasterContainerBaseVolumeMounts(jenkins *v1alpha2.Jenkins) []core
...
+		if env.Name == jenkinsHomeEnvVar.Name {
+			jenkinsHomeEnvVarExists = true
+			env.Value = jenkinsHomeEnvVar.Value
+		}
+	}
+
+	if !jenkinsHomeEnvVarExists {
+		envs = append(envs, jenkinsHomeEnvVar)
+	}
+
```
Comment (tomaszsek) [inline pkg/controller/jenkins/configuration/base/resources/pod.go]:
```
	jenkinsContainer := jenkins.Spec.Master.Containers[0]

	envs := GetJenkinsMasterContainerBaseEnvs(jenkins)
	envs = append(envs, jenkinsContainer.Env...)

	jenkinsHomeEnvVarExists := false
	for _, env := range jenkinsContainer.Env {
		if env.Name == jenkinsHomeEnvVar.Name {
			jenkinsHomeEnvVarExists = true
			break
		}
	}

	if !jenkinsHomeEnvVarExists {
		envs = append(envs, corev1.EnvVar{
		    Name:  "JENKINS_HOME",
		    Value: getJenkinsHomePath(jenkins),
	        })
	}
```
small improvement :)

Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/resources/pod.go]: Thanks tomas :)

**Pull request merged**